model.py

```diff
@@ -10,19 +10,31 @@
 from lstms import SplitLSTM
 
 class SubModel(nn.Module):
-    def __init__(self, num_const, num_var, num_hidden, num_out, num_groups):
+    def __init__(self, num_const, num_var, num_hidden, num_out, num_groups, num_batch, independent):
         super(SubModel, self).__init__()
-        self.lstm = SplitLSTM(num_const, num_var, num_hidden, num_groups, independent=True) # TODO make independent default
-        self.fc = SplitLinear(0, num_hidden, num_out, num_groups) # TODO add independent parameter for consistency
+        self.num_hidden = num_hidden
+        self.num_groups = num_groups
+        self.num_batch = num_batch
+
+        self.lstm = SplitLSTM(num_const, num_var, num_hidden, num_groups, independent)
+        self.bn = nn.BatchNorm1d(num_groups)
+        self.fc = SplitLinear(0, num_hidden, num_out, num_groups, independent)
+
+        self.hidden = (torch.zeros(1, num_batch, num_hidden*num_groups),
+                       torch.zeros(1, num_batch, num_hidden*num_groups))
+                    
+    def reset_hidden(self):
+        self.hidden = (torch.zeros(1, self.num_batch, self.num_hidden*self.num_groups),
```

Comment:
It does not work on the GPU, probably because self.hidden is not initialized on the GPU.

Reply:
A more maintainable way of fixing this would be to save the hidden state in `SplitLSTM`. Then when you call `reset_hidden()`, you set `self.hidden` to `None`. If you pass a `hx` that is equal to `None` to `self.lstm`, the layer will initialize the hidden state on the correct device automatically. This makes more sense, because the hidden state is more a part of the LSTM than of the entire sub-model.

Reply:
Will do. The reason why I did it this way is because I regarded `SplitLSTM` as a replacement of `nn.LSTM`. As such, `hidden` would be specified in the module where you assign the LSTM layer normally.

Reply:
It is a shame that the hidden state needs to be detached from the computational graph every iteration. This means that backpropagation through time (BPTT) will only work for a maximum of `seq_len` time steps. But alas, hopefully by having `seq_len > 1`, the training efficiency will still be increased compared to not having a hidden state.

Reply:
I moved the hidden state to `SplitLSTM` and it works on GPU now (note: you still have to change device in forward of `mlps.py` to `cuda`). GPU is much slower compared to CPU though from what I've seen.